sdk/messaging/azservicebus/receiver.go

```diff
@@ -341,123 +353,128 @@ func (r *Receiver) receiveMessage(ctx context.Context, options *ReceiveMessagesO
 }
 
 func (r *Receiver) receiveMessagesImpl(ctx context.Context, maxMessages int, options *ReceiveMessagesOptions) ([]*ReceivedMessage, error) {
-	// There are three phases for this function:
-	// Phase 1. <receive, respecting user cancellation>
-	// Phase 2. <check error and exit if fatal>
-	//    NOTE: We don't exit here so we don't end up buffering messages internally that the
-	//    user isn't actually waiting for anymore. So we make sure that #3 runs if the
-	//    link is still valid.
-	// Phase 3. <drain the link and leave it in a good state>
-
 	var all []*ReceivedMessage
-	var cancel context.CancelFunc
-	needsDrain := true
-
-	defaultTimeAfterFirstMessage := 20 * time.Millisecond
-
-	if r.receiveMode == ReceiveModeReceiveAndDelete {
-		defaultTimeAfterFirstMessage = time.Second
-	}
 
 	var linksWithID *internal.LinksWithID
 
-	err := r.amqpLinks.Retry(ctx, "receiveMessages", func(ctx context.Context, lwid *internal.LinksWithID, args *utils.RetryFnArgs) error {
-		if args.LastErr != nil {
-			// we recovered succesfully (amqpLinks does it for us), we can reset our retry attempts.
-			// This fixes a potential problem where something like this happens:
-			// a. amqplink.Receive() returns an error
-			// b. we attempt recovery a few times and recover just in the nick of time at attempt #3.
-			// d. amqplink.Receive() blocks (no messages in queue, for instance)
-			//   <years pass>
-			// e. amqpLink.Receive() returns an error
-			//
-			// We don't want to the # of attempts to already be '3' when we get to
-			// step 5, so we reset here so the next set of retry attempts starts fresh.
-			args.ResetAttempts()
-		}
-
+	err := r.amqpLinks.Retry(ctx, "receiveMessages.getlinks", func(ctx context.Context, lwid *internal.LinksWithID, args *utils.RetryFnArgs) error {
```

Comment:
What exactly is being retried here?

Reply:
I have two kinds of retry "primitives". The 'non'amqpLinks' version is just a Retry() like we've got in azcore without any special integration of anything.

The amqpLinks version specifically retries on creating the connection+links and understands that specific scenario enough that, by the time it succeeds and calls your callback, it'll definitely have acquired and creating everything. So after this set of retries is complete the links should be usable.